R/estimate_cox.R

```diff
@@ -0,0 +1,181 @@
+#' visR::estimate_cox(data = adtte, equation = c( "SEX", "survival::cluster(USUBJID)"),
+#'   tier = "breslow", conf.int = 0.9)
+#'
+#' ## Example working with non CDISC data
+#' head(survival::veteran)
+#'
+#' # convert time and censoring data to ADaM variables
+#' # convert censoring status to CDISC principles
+#' veteran_adam <- survival::veteran %>%
+#'  dplyr::mutate(AVAL = time,
+#'                CNSR = dplyr::if_else(status == 1, 0, 1)
+#'  )
+#'
+#' visR::estimate_cox(data = veteran_adam, equation = c("trt", "diagtime"))
+
+
+
+estimate_cox <- function(
+  data = NULL
+  ,equation = NULL
+  ,CNSR = "CNSR"
+  ,AVAL = "AVAL"
+  ,conf.int = 0.95
+  ,...
+){
+
+  # Capture input to validate user input for data argument -----------------
+
+  ## Get actual data name as symbol
+  ### Magrittre pipe returns "." which inactivates recalls to survfit in downstream functions
+  ### map passes .x as cox_ccall$data
+  ### df: catch expressions that represent base R subsets
+  cox_call <- as.list(match.call())
+  dots <- list(...)
+  df_expr <- cox_call[["data"]]
+
+  ## Validate `data` and capture data name
+
+  if (is.null(data)) stop(paste0("data can't be NULL."))
+
+  if (base::length(base::deparse(cox_call[["data"]])) == 1 && base::deparse(cox_call[["data"]]) %in% c(".", ".x")) {
+    df <- the_lhs()
+    cox_call[["data"]] <- as.symbol(df)
+  } else {
+    df <- as.character(sub("\\[.*$", "", deparse(df_expr))[1])
+  }
+
+  if (!(inherits(data, "data.frame") | inherits(data, "tibble") | inherits(data, "data.table"))) {
+    stop(paste0("data can be of class `data.frame` or `tibble` or `data.table`."))
+  }
+
+  data <- as.data.frame(data)
+
+  # Validate columns --------------------------------------------------------
+
+  reqcols <- c(equation, CNSR, AVAL)
+  ss <- c("strata", "tt", "frailty", "ridge", "pspline", "cluster", "I", "factor")
+
+  column_names <- gsub("\\([^()]*\\)", "", reqcols) 
+  column_names <- column_names[!column_names %in% ss]
+  column_names_special <- gsub("\\(([^()]*)\\)|.", "\\1", reqcols, perl=T)
+  column_names_special <- column_names_special[column_names_special!=""]
+  
+  reqcols_fin <- c(column_names, column_names_special)
+ 
+  # Make sure all variables are in dataset
+  if (! all(reqcols_fin %in% colnames(data))){
+    stop(paste0("Following columns are missing from `data`: ", paste(setdiff(reqcols_fin, colnames(data)), collapse = " "), "."))
+  }
+
+  if (!is.numeric(data[[AVAL]])) {
+    stop("Analysis variable (AVAL) is not numeric.")
+  }
+
+  if (!is.numeric(data[[CNSR]])) {
+    stop("Censor variable (CNSR) is not numeric.")
+  }
+
+
+  # Remove NA from the analysis --------------------------------------------
+
+  data <- as.data.frame(data) %>%
+    tidyr::drop_na(AVAL, CNSR)
+
+
+  # Ensure the presence of at least one strata -----------------------------
+
+  if (is.null(equation)) {
+    equation <- "1"
+  } else {
+    equation <- paste(equation, collapse = " + ")
```

Comment:
For the equation, I would let the user fill it out completely. I found it very confusing to understand the examples untill I saw the function collapses it. 
For me it does not make sense for the cox. If I have to fill out an equation, I dont want to think about how it is interpreted by the downstream function. Therefore, I would remove the following code:

  } else {
    equation <- paste(equation, collapse = " + ")
  }

Reply:
Done